### PR TITLE
Added methods related to conversions

### PR DESCRIPTION
**Pull request by luciaems**

### CODE REVIEW

File: rest/src/main/java/com/abiquo/apiclient/templates/TemplatesApi.java [hunk 1]
```diff
@@ -180,4 +202,82 @@ public VirtualMachineTemplateDto createPersistent(final VirtualDatacenterDto vdc
     {
         return client.list(vmt.searchLink("tasks").getHref(), TasksDto.MEDIA_TYPE, TasksDto.class);
     }
+
+    public AcceptedRequestDto<String> createOrRestartConversion(
```
Comment (nacx) [inline rest/src/main/java/com/abiquo/apiclient/templates/TemplatesApi.java]: Split this method in two:
```java
createConversion(final VirtualMachineTemplateDto template, final String targetFormat);
restartConversion(final ConversionDto conversion);
```

File: rest/src/main/java/com/abiquo/apiclient/templates/TemplatesApi.java [hunk 2]
```diff
@@ -180,4 +202,82 @@ public VirtualMachineTemplateDto createPersistent(final VirtualDatacenterDto vdc
...
+    }
+
+    public VirtualMachineTemplateDto downloadTemplateToRepository(
+        final DatacenterRepositoryDto repository, final TemplateDefinitionDto templateDefinition,
+        final int polling, final int timeout, final TimeUnit time)
+    {
+
+        VirtualMachineTemplateRequestDto templateDefinitionReq =
+            new VirtualMachineTemplateRequestDto();
+        templateDefinitionReq.addLink(create("templateDefinition", templateDefinition.getEditLink()
```
Comment (nacx) [inline rest/src/main/java/com/abiquo/apiclient/templates/TemplatesApi.java]: To "copy" a link changing the "rel" (which is a common thing) you can better use the [Links#withRel](https://github.com/abiquo/api-java-client/blob/templateConversion/rest/src/main/java/com/abiquo/apiclient/domain/Links.java#L30-L35) method.

File: rest/src/main/java/com/abiquo/apiclient/domain/ApiPath.java [hunk 3]
```diff
@@ -23,6 +23,8 @@
 
     public static final String LOADLEVELRULES_URL = "/admin/rules/machineLoadLevel";
 
+    public static final String PUBLIC_CLOUD_REGIONS_URL = "admin/publiccloudregions";
```
Comment (nacx) [inline rest/src/main/java/com/abiquo/apiclient/domain/ApiPath.java]: Change it to start with a `/` like the other URLs.

File: rest/src/test/java/com/abiquo/apiclient/infrastructure/InfrastructureApiTest.java [hunk 4]
```diff
@@ -698,4 +703,79 @@ public void testCreateMachineLoadLevelRule() throws Exception
...
+        MockResponse response =
+            new MockResponse().setHeader("Content-Type",
+                PublicCloudRegionsDto.SHORT_MEDIA_TYPE_JSON).setBody(
+                payloadFromResource("pcrs.json"));
+
+        server.enqueue(response);
+        server.play();
+
+        newApiClient().getInfrastructureApi().listPublicCloudRegions(
+            PublicCloudRegionListOptions.builder().limit(1).build());
```
Comment (nacx) [inline rest/src/test/java/com/abiquo/apiclient/infrastructure/InfrastructureApiTest.java]: Better use the options specific to this class instead of the ones of the parent.

File: rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java [hunk 5]
```diff
@@ -97,6 +102,59 @@ public void testListTemplatesWithOptions() throws Exception
...
+
+        DatacenterRepositoryDto repository = new DatacenterRepositoryDto();
+        RESTLink link =
+            new RESTLink("virtualmachinetemplates",
+                "/admin/enterprises/1/datacenterrepositories/2/virtualmachinetemplates");
+        link.setType(VirtualMachineTemplatesDto.SHORT_MEDIA_TYPE_JSON);
+        repository.addLink(link);
+
+        newApiClient().getTemplatesApi().listTemplates(repository,
+            TemplateListOptions.builder().limit(1).build());
```
Comment (nacx) [inline rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java]: Same here, better use the `TemplateListOptions` options.

File: rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java [hunk 6]
```diff
@@ -534,4 +592,195 @@ public void testGetVirtualMachineTemplateTasks() throws Exception
...
+            "/admin/enterprises/1/datacenterrepositories/1/virtualmachinetemplates/1"));
+        newApiClient().getTemplatesApi().createOrRestartConversion(vmt, "VMDK_FLAT", null);
+
+        RecordedRequest request = server.takeRequest();
+        assertRequest(request, "PUT",
+            "/admin/enterprises/1/datacenterrepositories/1/virtualmachinetemplates/1/conversions/VMDK_FLAT");
+        assertAccept(request, AcceptedRequestDto.SHORT_MEDIA_TYPE_JSON,
+            SingleResourceTransportDto.API_VERSION);
+
+    }
```
Comment (nacx) [inline rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java]: There is missing a test when the "conversion" parameter is not null, but let's add it when you split the method in two.

File: rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java [hunk 7]
```diff
@@ -534,4 +592,195 @@ public void testGetVirtualMachineTemplateTasks() throws Exception
...
+        VirtualMachineTemplateDto vmt = new VirtualMachineTemplateDto();
+        vmt.addLink(new RESTLink("edit",
+            "/admin/enterprises/1/datacenterrepositories/1/virtualmachinetemplates/1"));
+        newApiClient().getTemplatesApi().createOrRestartConversion(vmt, "VMDK_FLAT", null);
+
+        RecordedRequest request = server.takeRequest();
+        assertRequest(request, "PUT",
+            "/admin/enterprises/1/datacenterrepositories/1/virtualmachinetemplates/1/conversions/VMDK_FLAT");
+        assertAccept(request, AcceptedRequestDto.SHORT_MEDIA_TYPE_JSON,
+            SingleResourceTransportDto.API_VERSION);
```
Comment (nacx) [inline rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java]: The PUT request will set the content-type. You should check it too.

File: rest/src/main/java/com/abiquo/apiclient/RestClient.java [hunk 8]
```diff
@@ -310,6 +310,25 @@ public void delete(final String uri)
...
+                    .addHeader(HttpHeaders.AUTHORIZATION, authHeader)
+                    .addHeader(HttpHeaders.ACCEPT, withVersion(accept)).post(requestBody).build();
+
+            return execute(request, returnClass);
+        }
+        catch (IOException ex)
+        {
+            throw Throwables.propagate(ex);
+        }
+    }
```
Comment (nacx) [inline rest/src/main/java/com/abiquo/apiclient/RestClient.java]: All post methods that have a body should require the Content-Type header describing the body, so this method should be removed.
The method that uses it, is expected to use the "text/plain" content type, so you can explicitly use that and get rid of this method.

File: rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java [hunk 9]
```diff
@@ -534,4 +592,195 @@ public void testGetVirtualMachineTemplateTasks() throws Exception
...
+
+        EnterpriseDto enterprise = new EnterpriseDto();
+        enterprise.setId(234);
+
+        newApiClient().getTemplatesApi().createTemplateDefinitionList(enterprise, "urlRepo");
+
+        RecordedRequest request = server.takeRequest();
+        assertRequest(request, "POST", "/admin/enterprises/234/appslib/templateDefinitionLists");
+        assertAccept(request, TemplateDefinitionListDto.SHORT_MEDIA_TYPE_JSON,
+            SingleResourceTransportDto.API_VERSION);
```
Comment (nacx) [inline rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java]: You should also verify the content-type once you set it ti text/plain.

File: rest/src/main/java/com/abiquo/apiclient/templates/TemplatesApi.java [hunk 10]
```diff
@@ -180,4 +202,82 @@ public VirtualMachineTemplateDto createPersistent(final VirtualDatacenterDto vdc
...
+
+    }
+
+    public Iterable<TaskDto> listConversionTasks(final VirtualMachineTemplateDto vmt,
+        final String diskFormatType)
+    {
+
+        return client.list(
+            String.format("%s/%s/tasks", vmt.searchLink("conversions").getHref(), diskFormatType),
+            TasksDto.MEDIA_TYPE, TasksDto.class);
```
Comment (nacx) [inline rest/src/main/java/com/abiquo/apiclient/templates/TemplatesApi.java]: Does the ConversionDto object have a "tasks" link? If it has one, it would be better that this method received a ConversionDto as a parameter instead of the current ones.

File: rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java [hunk 11]
```diff
@@ -534,4 +592,195 @@ public void testGetVirtualMachineTemplateTasks() throws Exception
...
+        assertRequest(
+            request,
+            "GET",
+            "/admin/enterprises/1/datacenterrepositories/1/virtualmachinetemplates/1/conversions/VMDK_FLAT/tasks");
+        assertAccept(request, TasksDto.SHORT_MEDIA_TYPE_JSON,
+            SingleResourceTransportDto.API_VERSION);
+
+    }
+
+    public void testListConversion() throws Exception
```
Comment (nacx) [inline rest/src/test/java/com/abiquo/apiclient/templates/TemplatesApiTest.java]: [minor] rename to testListConversions :)

**Comment by nacx** [comment]
+1
